### ISSUE TITLE
Bug: Types from types/node are detected as error types

### ISSUE THREAD

**Issue by anonrig** [typescript-eslint/typescript-eslint#10635]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

eslint-plugin

### Repro Code

Example: https://github.com/cloudflare/workerd/pull/3315

### ESLint Config

```javascript
import eslint from "@eslint/js";
import tseslint from "typescript-eslint";
import prettierPlugin from "eslint-plugin-prettier/recommended";

/**
 * @param {string} tsconfigRootDir
 * @returns {FlatConfig.ConfigArray}
 */
export function baseConfig({ tsconfigRootDir }) {
  return tseslint.config(
    eslint.configs.recommended,
    ...tseslint.configs.strictTypeChecked,
    {
      languageOptions: {
        parserOptions: {
          ecmaVersion: "latest",
          sourceType: "module",
          projectService: true,
          tsconfigRootDir,
        },
      },
      rules: {
        "@typescript-eslint/explicit-function-return-type": "error",
        "@typescript-eslint/explicit-member-accessibility": "error",
        "@typescript-eslint/explicit-module-boundary-types": "error",
        "@typescript-eslint/no-require-imports": "error",
        "@typescript-eslint/prefer-enum-initializers": "error",
        "@typescript-eslint/type-annotation-spacing": "error",
        "@typescript-eslint/restrict-template-expressions": "warn",
        "@typescript-eslint/no-non-null-assertion": "warn",
        "@typescript-eslint/no-extraneous-class": "off",
        "@typescript-eslint/unified-signatures": "off",
        "@typescript-eslint/no-unused-vars": [
          "error",
          {
            args: "all",
            argsIgnorePattern: "^_",
            caughtErrors: "all",
            caughtErrorsIgnorePattern: "^_",
            destructuredArrayIgnorePattern: "^_",
            varsIgnorePattern: "^_",
            ignoreRestSiblings: true,
          },
        ],
      },
    },
    prettierPlugin,
    {
      files: ['**/*.js', '**/*.mjs', '**/*.cjs'],
      ...tseslint.configs.disableTypeChecked,
    },
  )
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "lib": ["ESNext", "dom"],
    "alwaysStrict": true,
    "strict": true,
    "allowJs": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "exactOptionalPropertyTypes": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noPropertyAccessFromIndexSignature": false,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "strictNullChecks": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "declaration": true,
    "types": ["@types/node"],
    "paths": {
      "node:*": ["./*"],
      "node:assert/*": ["./*"],
      "node:dns/*": ["./*"],
      "node:util/*": ["./*"],
      "node:path/*": ["./*"],
      "node:stream/*": ["./*"],
      "node-internal:*": ["./internal/*"],
      "cloudflare-internal:sockets": ["./internal/sockets.d.ts"],
      "cloudflare-internal:workers": ["./internal/workers.d.ts"],
      "workerd:compatibility-flags": ["./internal/compatibility-flags.d.ts"]
    }
  },
  "exclude": []
}
```

### Expected Result

Detect types from types/node

### Actual Result

Logs: https://github.com/cloudflare/workerd/actions/runs/12699380285/job/35400099540?pr=3315

All types from types/node are detected as `error` type:

```
   128:20  error    'SocketType' is an 'error' type that acts as 'any' and overrides all other types in this intersection type                                                                                                                      @typescript-eslint/no-redundant-type-constituents
   133:19  error    'any' overrides all other types in this union type                                                                                                                                                                              @typescript-eslint/no-redundant-type-constituents
   144:15  error    'any' overrides all other types in this union type                                                                                                                                                                              @typescript-eslint/no-redundant-type-constituents
   176:22  error    'AddressInfo' is an 'error' type that acts as 'any' and overrides all other types in this union type
```

### Versions

```
{
  "dependencies": {
    "typescript": "5.5.4"
  },
  "devDependencies": {
    "@eslint/js": "^9.10.0",
    "@types/eslint__js": "^8.42.3",
    "@types/node": "^20.16.5",
    "eslint": "^9.10.0",
    "eslint-plugin-import": "^2.30.0",
    "eslint-plugin-prettier": "^5.2.1",
    "typescript-eslint": "^8.4.0"
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
> `projectService: true`
> ...
>
> ```plaintext 
>   128:20  error    'SocketType' is an 'error' type that acts as 'any' and overrides all other types in this intersection type                                                                                                                      @typescript-eslint/no-redundant-type-constituents
> ```

This is _probably_ not a problem with typescript-eslint. `projectService` just pulls in whatever type information TypeScript computes by default. The code paths internally are almost identical to what editors like VS Code use. If you're getting this, then it's probably that there's a bug somewhere in how you've configured TypeScript services. That might be the lint config, typescript-eslint, or TypeScript itself.

But, many -likely most- of our users include `@types/node` in their setups. We do it in this repository and I do it in all of my personal ones. So it's _probably_ an issue with your lint configuration.

> ```plaintext
> "@typescript-eslint/type-annotation-spacing": "error",
> ```

When I try to run this lint config, I get:

```plaintext
TypeError: Key "rules": Key "@typescript-eslint/type-annotation-spacing": Could not find "type-annotation-spacing" in plugin "@typescript-eslint".
```

Because that rule hasn't existed in typescript-eslint since major version 7: #8074 -> #8833.

I tried cloning https://github.com/cloudflare/workerd/pull/3315 locally but there's no ESLint config in the repro and I don't have any Rust tooling installed locally.

Could you please post a reproduction that I can quickly clone and get working on my local computer?

**Comment by anonrig** [comment]
Thanks. I've removed the invalid rule on https://github.com/cloudflare/workerd/pull/3320, once it lands I'l update the original PR to see if there is any change.
